compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUPromoteMatmulToFitMMA.cpp

```diff
@@ -0,0 +1,167 @@
+#define GEN_PASS_DECL_LLVMGPUPROMOTEMATMULTOFITMMA
+#include "iree/compiler/Codegen/LLVMGPU/Passes.h.inc"
+namespace {
+
+class LLVMGPUPromoteMatmulToFitMMAPass
+    : public LLVMGPUPromoteMatmulToFitMMABase<
+          LLVMGPUPromoteMatmulToFitMMAPass> {
+public:
+  explicit LLVMGPUPromoteMatmulToFitMMAPass(
+      const LLVMGPUMatmulPadOption &option) {
+    this->targetDimensions.setValue(option);
+  }
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<tensor::TensorDialect, linalg::LinalgDialect>();
+  }
+
+  void pad(RewriterBase &rewriter, linalg::LinalgOp op,
+           utils::IteratorType targetIterType, bool nofold) const {
+    LLVM_DEBUG(llvm::dbgs() << "candidate: " << op << "\n");
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointAfter(op);
+
+    SmallVector<int64_t> paddingDims;
+    for (auto [index, iterType] : llvm::enumerate(op.getIteratorTypesArray())) {
+      if (iterType == targetIterType) {
+        paddingDims.push_back(index);
+      }
+    }
+
+    SmallVector<bool> packPaddings(op.getNumDpsInputs(), nofold);
+
+    // One is enough because they will essentially be padded to corresponding
+    // tile sizes, which should be multiple of MMA shapes.
+    SmallVector<int64_t> padToMultipleOf(paddingDims.size(), 1);
+    SmallVector<Attribute> paddingValueAttributes;
+    for (auto &operand : op->getOpOperands()) {
+      auto elemType = getElementTypeOrSelf(operand.get().getType());
+      paddingValueAttributes.push_back(rewriter.getZeroAttr(elemType));
+    }
+
+    auto options =
+        linalg::LinalgPaddingOptions()
+            .setPaddingDimensions(paddingDims)
+            .setPaddingValues(paddingValueAttributes)
+            .setPadToMultipleOf(padToMultipleOf)
+            .setPackPaddings(packPaddings)
+            .setCopyBackOp(linalg::LinalgPaddingOptions::CopyBackOp::None);
+
+    FailureOr<linalg::LinalgOp> result =
+        linalg::padAndHoistLinalgOp(rewriter, op, options);
+    if (failed(result)) {
+      LLVM_DEBUG(llvm::dbgs() << "failed to pad op " << op << "\n");
+    }
+  }
+
+  void runOnOperation() override {
+    MLIRContext *ctx = &getContext();
+    auto funcOp = getOperation();
+
+    // Preserve the innermost tensor.pad ops (i.e., pad for reduction dims), so
+    // we can kick canonicalization patterns to fold outer tensor.pad ops away.
+    bool nofold = false;
+    utils::IteratorType targetIterType = utils::IteratorType::parallel;
+    switch (targetDimensions) {
+    case LLVMGPUMatmulPadOption::ParallelDims:
+      LLVM_DEBUG(llvm::dbgs() << "padding parallel dims\n");
+      targetIterType = utils::IteratorType::parallel;
+      nofold = false;
+      break;
+    case LLVMGPUMatmulPadOption::ReductionDims:
+      LLVM_DEBUG(llvm::dbgs() << "padding reduction dims\n");
+      targetIterType = utils::IteratorType::reduction;
+      nofold = true;
+      break;
+    default: // Unreachable.
+      assert(false);
```

Comment:
llvm_unreachable?

Reply:
This is an interesting comment, and bring me to find @bjacob 's old post: https://github.com/iree-org/iree/pull/8521

I re-studied the thread and think that we want to avoid `llvm_unreachable` in this case. It could potentially lead to UB. The only benefit we could have is small performance win, and it rarely happens. IREE project is following the style, so I'd like to keep `assert(false)` here.

Reply:
I see, thanks for pointing me to the interesting discussion, makes sense to leave as an assert then.